nexus/reconfigurator/planning/src/planner.rs

```diff
@@ -4337,4 +4474,487 @@ pub(crate) mod test {
+            &logctx.log,
+            rng.next_system_rng(),
+        )
+        .build();
+        verify_blueprint(&blueprint1);
+
+        // We should start with no specified TUF repo and nothing to do.
+        assert!(example.input.tuf_repo().is_none());
+        // assert_planning_makes_no_changes(
+        //     &logctx.log,
+        //     &blueprint1,
+        //     &example.input,
+        //     &example.collection,
+        //     TEST_NAME,
+        // );
+
+        // All zones should be sourced from the install dataset by default.
+        assert!(
+            blueprint1
+                .all_omicron_zones(BlueprintZoneDisposition::is_in_service)
+                .all(|(_, z)| matches!(
+                    z.image_source,
+                    BlueprintZoneImageSource::InstallDataset
+                ))
+        );
+
+        // Manually specify a trivial TUF repo.
+        let mut input_builder = example.input.clone().into_builder();
+        input_builder.policy_mut().tuf_repo = Some(TufRepoDescription {
+            repo: TufRepoMeta {
+                hash: ArtifactHash([0; 32]),
+                targets_role_version: 0,
+                valid_until: Utc::now(),
+                system_version: Version::new(0, 0, 0),
+                file_name: String::from(""),
+            },
+            artifacts: vec![],
+        });
+        let input = input_builder.build();
+        let blueprint2 = Planner::new_based_on(
+            log.clone(),
+            &blueprint1,
+            &input,
+            "test_blueprint2",
+            &example.collection,
+        )
+        .expect("can't create planner")
+        .with_rng(PlannerRng::from_seed((TEST_NAME, "bp2")))
+        .plan()
+        .expect("plan for trivial TUF repo");
+
+        // All zones should still be sourced from the install dataset.
+        assert!(
+            blueprint2
+                .all_omicron_zones(BlueprintZoneDisposition::is_in_service)
+                .all(|(_, z)| matches!(
+                    z.image_source,
+                    BlueprintZoneImageSource::InstallDataset
+                ))
+        );
+
+        // Manually specify a TUF repo with fake zone images for Crucible Pantry
+        // and Nexus. Only the name and kind of the artifacts matter. The Nexus
+        // artifact is only there to make sure the planner *doesn't* use it.
+        let mut input_builder = input.into_builder();
+        let version = ArtifactVersion::new_static("1.0.0-freeform")
+            .expect("can't parse artifact version");
+        let fake_hash = ArtifactHash([0; 32]);
+        let image_source = BlueprintZoneImageSource::Artifact {
+            version: BlueprintZoneImageVersion::Available {
+                version: version.clone(),
+            },
+            hash: fake_hash,
+        };
+        let artifacts = vec![
+            TufArtifactMeta {
+                id: ArtifactId {
+                    name: String::from("crucible-pantry-zone"),
```

Comment:
Nit - maybe

```suggestion
                    name: ZoneKind::CruciblePantry.artifact_name().to_string(),
```

instead (and similar for other names below)? Slightly longer, but avoids any concerns about typos or what the string should be.

Reply:
Thank you, fixed in 88733d257 for this test and the next by refactoring its little macro, now called `fake_zone_artifact`.